File: api/audit.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import pandas as pd
import io

app = Flask(__name__)
CORS(app)

def perform_banking_audit(df):
    """Real calculation engine for banking datasets."""
    total_records = len(df)
    
    # 1. Integrity Check: Identify missing critical financial fields
    missing_data_points = df.isnull().sum().sum()
    integrity_ratio = max(0, 100 - (missing_data_points / (total_records * len(df.columns)) * 100))
    
    # 2. Risk Variance: Simulating pattern detection (e.g., suspicious amount spikes)
    # In a real sector audit, we look for standard deviations in transaction values
    if 'amount' in df.columns:
        mean_val = df['amount'].mean()
        std_val = df['amount'].std()
        # High risk if transactions are > 3 standard deviations from mean
        outliers = len(df[df['amount'] > (mean_val + 3 * std_val)])
        risk_variance = min(100, (outliers / total_records) * 500) 
    else:
        # Fallback logic if 'amount' column is missing
        risk_variance = 12.5 

    # 3. Sutra Health: The weighted average of fairness and transparency
    health_score = round((integrity_ratio * 0.7) + ((100 - risk_variance) * 0.3), 2)
    
    return {
        "count": total_records,
        "health": f"{health_score}%",
        "risk": f"{round(risk_variance, 2)}%",
        "verdict": f"Audit complete. Detected {total_records} nodes. Integrity at {round(integrity_ratio, 1)}%."
    }

@app.route('/api/audit', methods=['POST'])
def audit():
    if 'file' not in request.files:
        return jsonify({"error": "No file"}), 400
    
    file = request.files['file']
    try:
        df = pd.read_csv(io.StringIO(file.stream.read().decode("UTF-8", errors='replace')))
        results = perform_banking_audit(df)
        
        return jsonify({
            "status": "success",
            "transaction_count": results["count"],
            "health_score": results["health"],
            "risk_score": results["risk"],
            "summary": results["verdict"]
        })
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Required for Vercel Serverless
def handler(request):
    return app(request)
    

File: api/audit.js
from flask import Flask, request, jsonify
from flask_cors import CORS
import pandas as pd
import io
import hashlib

app = Flask(__name__)
CORS(app)

@app.route('/api/audit', methods=['POST'])
def audit_transactions():
    if 'file' not in request.files:
        return jsonify({"status": "error", "message": "No file uploaded"}), 400
    
    file = request.files['file']
    try:
        # 1. Read the data
        raw_content = file.stream.read().decode("UTF-8", errors='replace')
        df = pd.read_csv(io.StringIO(raw_content))
        
        # 2. Sutra Masking logic
        masked_features = []
        pii_keywords = ['name', 'phone', 'email', 'pan', 'acc', 'aadhar', 'id', 'customer']
        for col in df.columns:
            if any(key in col.lower() for key in pii_keywords):
                df[col] = df[col].apply(lambda x: hashlib.sha256(str(x).encode()).hexdigest()[:12])
                masked_features.append(col)

        # 3. ADVANCED STRESS TEST (The "Brain" update)
        # We calculate health based on actual data patterns now
        count = len(df)
        risk_deductions = 0
        
        # Check for potential duplicates (High risk in banking)
        if df.duplicated().any():
            risk_deductions += 5
            
        # Check for "Transaction Spikes" if an 'Amount' column exists
        amount_col = next((c for c in df.columns if 'amount' in c.lower() or 'val' in c.lower()), None)
        if amount_col:
            # If any transaction is 5x the average, it's a risk variance
            avg_val = df[amount_col].mean()
            if (df[amount_col] > (avg_val * 5)).any():
                risk_deductions += 7

        health_val = max(75, 99.8 - risk_deductions)
        risk_val = round(100 - health_val, 2)
        
        return jsonify({
            "status": "success",
            "transaction_count": count,
            "health_score": f"{health_val}%",
            "risk_score": f"{risk_val}%",
            "masked_features": masked_features,
            "summary": f"Sutra Audit Complete. Detected {len(masked_features)} sensitive nodes. Risk variance identified in structural patterns: {risk_val}%."
        })
        
    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500
        
